app/src/hid_keyboard.c

```diff
@@ -0,0 +1,347 @@
+#include "util/log.h"
+#include "hid_keyboard.h"
+
+/**
+ * For HID over AOAv2, we only need report descriptor.
+ * Normally a basic HID keyboard uses 8 bytes,
+ * `Modifier Reserved Key Key Key Key Key Key`.
+ * See Appendix B.1 Protocol 1 (Keyboard) and
+ * Appendix C: Keyboard Implementation in
+ * <https://www.usb.org/sites/default/files/hid1_11.pdf>.
+ * But if we want to support media keys on keyboard,
+ * we need to use two reports,
+ * report id 1 and usage page key codes for basic keyboard,
+ * report id 2 and usage page consumer for media keys.
+ * See 8. Report Protocol in
+ * <https://www.usb.org/sites/default/files/hid1_11.pdf>.
+ * The former byte is item type prefix, we only use short items here.
+ * For how to calculate an item, read 6.2.2 Report Descriptor in
+ * <https://www.usb.org/sites/default/files/hid1_11.pdf>.
+ * For Consumer Page tags, see 15 Consumer Page (0x0C) in
+ * <https://usb.org/sites/default/files/hut1_2.pdf>.
+ */
+/**
+ * You can dump your device's report descriptor with
+ * `sudo usbhid-dump -m vid:pid -e descriptor`.
+ * Change `vid:pid` to your device's vendor ID and product ID.
+ */
+unsigned char kb_report_desc_buffer[]  = {
+    // Usage Page (Generic Desktop)
+    0x05, 0x01,
+    // Usage (Keyboard)
+    0x09, 0x06,
+
+    // Collection (Application)
+    0xA1, 0x01,
+    // Report ID (1)
+    0x85, 0x01,
+
+    // Usage Page (Keyboard)
+    0x05, 0x07,
+    // Usage Minimum (224)
+    0x19, 0xE0,
+     // Usage Maximum (231)
+    0x29, 0xE7,
+    // Logical Minimum (0)
+    0x15, 0x00,
+    // Logical Maximum (1)
+    0x25, 0x01,
+    // Report Size (1)
+    0x75, 0x01,
+    // Report Count (8)
+    0x95, 0x08,
+    // Input (Data, Variable, Absolute): Modifier byte
+    0x81, 0x02,
+
+    // Report Size (8)
+    0x75, 0x08,
+    // Report Count (1)
+    0x95, 0x01,
+    // Input (Constant): Reserved byte
+    0x81, 0x01,
+
+    // Usage Page (LEDs)
+    0x05, 0x08,
+    // Usage Minimum (1)
+    0x19, 0x01,
+    // Usage Maximum (5)
+    0x29, 0x05,
+    // Report Size (1)
+    0x75, 0x01,
+    // Report Count (5)
+    0x95, 0x05,
+    // Output (Data, Variable, Absolute): LED report
+    0x91, 0x02,
+
+    // Report Size (3)
+    0x75, 0x03,
+    // Report Count (1)
+    0x95, 0x01,
+    // Output (Constant): LED report padding
+    0x91, 0x01,
```

Comment:
Are these 18 bytes about LEDs necessary in the case of scrcpy?

They appear in [hid1_11.pdf](https://www.usb.org/sites/default/files/hid1_11.pdf) n Appendinx B §`B.1 Protocol 1 (Keyboard)`, but it's not clear to me if it is just an example or if it must contain these items.

The Appendix B precises:

> Devices may append additional data to these boot reports, but the first 8 bytes of keyboard reports and the first 3 bytes of mouse reports must conform to the format defined by the Boot Report descriptor in order for the data to be correctly interpreted by the BIOS.

So I think we could remove these bytes safely. At least it works on my device.

Note: On the other hand, if we start mirroring with <kbd>CAPS LOCK</kbd> (or <kbd>NUM LOCK</kbd>) on, then the state on the device is reversed. Maybe we could synchronize states? However, it could be a bit tricky, since several scrcpy windows mirroring different devices might not be in the same state, so it would require to synchronize on window focus, etc. I'm not sure it is worth it. (I'm just thinking out loud.)

Reply:
> Are these 18 bytes about LEDs necessary in the case of scrcpy?

I suggest to keep them because they are harmless, we never use them, but every keyboards have LEDs, we'd better behave more like a real keyboard.

Remove those LEDs should work, but maybe some systems require them, I am not sure. So I suggest to keep those harmless definations.

Nearly every examples on Internet have those lines, so if some one wants to modify those code, they can compare those with examples line by line if we keep them.

> Note: On the other hand, if we start mirroring with CAPS LOCK (or NUM LOCK) on, then the state on the device is reversed. Maybe we could synchronize states? However, it could be a bit tricky, since several scrcpy windows mirroring different devices might not be in the same state, so it would require to synchronize on window focus, etc. I'm not sure it is worth it. (I'm just thinking out loud.)

I think user is responsible for sync those state, if we do sync in code, a lot of hack is needed (those LEDs are control by system so you need permission to modify keyboard state), grabbing focus may be not allowed on some platform (Wayland I think), so the simplest way is let user handle it to start scrcpy with CAPS LOCK off.

Reply:
Another way to sync Caps Locks may be that when we start scrcpy, we detect the state of those locks, and if they are on, we send HID events for them before handling user actions. 😄(However I am not sure to do it, maybe we can implement it future.)